Refresh APM Troubleshooting Guide (#24686)
DataDog/documentation: content/en/tracing/troubleshooting/_index.md

| Other products <br> <i> (Dashboard, Notebook etc.)</i> | Spans from custom retention filters + Datadog intelligent filter |

{{% /collapse-content %}}

## Trace metrics

This section covers troubleshooting discrepancies and inconsistencies with trace metrics.

{{% collapse-content title="Trace metrics and custom span-based metrics have different values" level="h4" %}}

Trace metrics and custom span-based metrics can have different values because they are calculated based on different datasets:

- [Trace metrics][24] are calculated based on 100% of the application's traffic, regardless of your [trace ingestion sampling][25] configuration. The trace metrics namespace follows this format: `trace.<SPAN_NAME>.<METRIC_SUFFIX>`.
- [Custom span-based metrics][26] are generated based on your ingested spans, which depend on your [trace ingestion sampling][25]. For example, if you are ingesting 50% of your traces, your custom span-based metrics are based on the 50% ingested spans.

To ensure that your trace metrics and custom span-based metrics have the same value, configure a 100% ingestion rate for your application or service.

> <div class="alert alert-info">Metric names must follow the <a href="/metrics/#naming-metrics">metric naming convention</a>. Metric names that start with <code>trace.*</code> are not permitted and are not saved.</div>

Author: Good catch. I updated it to point here now: /metrics/custom_metrics/#naming-custom-metrics

{{% /collapse-content %}}

## Security

This section covers approaches for addressing security concerns in APM, including protecting sensitive data and managing traffic.

{{% collapse-content title="Modifying, discarding, or obfuscating spans" level="h4" %}}

There are several configuration options available to scrub sensitive data or discard traces corresponding to health checks or other unwanted traffic that can be configured within the Datadog Agent, or in some languages the Tracing Client. For details on the options available, see [Security and Agent Customization][11]. While this offers representative examples, if you require assistance applying these options to your environment, reach out to [Datadog Support][1].

{{% /collapse-content %}}

## Debugging and logging

This section explains how to use debug and startup logs to identify and resolve issues with your Datadog tracer.

{{% collapse-content title="Debug logs" level="h4" %}}

> To capture full details on the Datadog tracer, enable debug mode on your tracer by using the `DD_TRACE_DEBUG` environment variable. You might enable it for your own investigation or because Datadog support recommended it for triage purposes. However, don't leave debug mode always enabled because of the logging overhead it introduces.

Reviewer: ```suggestion
To capture full details on the Datadog tracer, enable debug mode on your tracer by using the `DD_TRACE_DEBUG` environment variable. You might enable it for your own investigation or if Datadog support has recommended it for triage purposes. However, be sure to disable debug logging when you are finished testing to avoid the logging overhead it introduces.
```

{{% collapse-content title="Open a Datadog Support ticket" level="h4" %}}

When you open a [support ticket][1], the Datadog support team may ask for the following types of information:

1. **Links to a trace or screenshots of the issue**: This helps reproduce your issues for troubleshooting purposes.

2. **Tracer startup logs**: Startup logs help identify tracer misconfiguration or communication issues between the tracer and the Datadog Agent. By comparing the tracer's configuration with the application or container settings, support teams can pinpoint improperly applied settings.

3. **Tracer debug logs**: Tracer debug logs provide deeper insights than startup logs, revealing:
   - Proper integration instrumentation during application traffic flow
   - Contents of spans created by the tracer
   - Connection errors when sending spans to the Agent

4. **Datadog Agent flare**: [Datadog Agent flares][12] enable you to see what is happening within the Datadog Agent, for example, if traces are being rejected or malformed. This does not help if traces are not reaching the Datadog Agent, but does help identify the source of an issue, or any metric discrepancies.

> 5. **A description of your environment**: Understanding your application's deployment configuration helps the Support team identify potential tracer-agent communication issues and spot misconfigurations. For complex problems, Support may request Kubernetes manifests, ECS task definitions, or similar deployment configuration files.

Reviewer: ```suggestion
5. **A description of your environment**: Understanding your application's deployment configuration helps the Support team identify potential tracer-Agent communication issues and identify misconfigurations. For complex problems, support may request Kubernetes manifests, ECS task definitions, or similar deployment configuration files.
```

6. **Custom tracing code**: Custom instrumentation, configuration, and adding span tags can significantly impact trace visualizations in Datadog.

> 7. **Version information**: Knowing what language, framework, Datadog Agent, and Datadog Tracer versions you are using allows Support to verify [Compatiblity Requirements][15], check for known issues, or recommend a version upgrades. For example:

Reviewer: ```suggestion
7. **Version information**: Knowing what language, framework, Datadog Agent, and Datadog tracer versions you are using allows Support to verify [Compatiblity Requirements][15], check for known issues, or recommend a version upgrades. For example:
```

Also don't think tracer is capitalized in this context ?
